modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C

```diff
@@ -0,0 +1,242 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org All rights reserved, see COPYRIGHT
+//* for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT Licensed
+//* under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdvectiveFluxCalculator.h"
+
+#include "libmesh/string_to_enum.h"
+#include "libmesh/mesh_tools.h"
+
+registerMooseObject("PorousFlowApp", AdvectiveFluxCalculator);
+
+template <>
+InputParameters
+validParams<AdvectiveFluxCalculator>()
+{
+  InputParameters params = validParams<ElementLoopUserObject>();
+  params.addClassDescription("Computes K_ij (a measure of advective flux from node i to node j) "
+                             "and R+ and R- (which quantify amount of antidiffusion to add) in the "
+                             "Kuzmin-Turek FEM-TVD multidimensional scheme");
+  params.addRequiredCoupledVar("u", "The variable that is being advected");
+  params.addRequiredParam<RealVectorValue>("velocity", "Velocity vector");
+  MooseEnum flux_limiter_type("MinMod VanLeer MC superbee None", "MinMod");
+  params.addParam<MooseEnum>("flux_limiter_type",
+                             flux_limiter_type,
+                             "Type of flux limiter to use.  'None' means that no antidiffusion "
+                             "will be added in the Kuzmin-Turek scheme");
+  // QUERY: do we need timestep_begin (see notes in timestepSetup method) or just nonlinear?
+  params.set<ExecFlagEnum>("execute_on", true) = {EXEC_TIMESTEP_BEGIN, EXEC_NONLINEAR};
+  return params;
+}
+
+AdvectiveFluxCalculator::AdvectiveFluxCalculator(const InputParameters & parameters)
+  : ElementLoopUserObject(parameters),
+    _velocity(getParam<RealVectorValue>("velocity")),
+    _u_nodal(getVar("u", 0)),
+    _u_var_num(coupled("u", 0)),
+    _phi(_assembly.fePhi<Real>(_u_nodal->feType())),
+    _grad_phi(_assembly.feGradPhi<Real>(_u_nodal->feType())),
+    _flux_limiter_type(getParam<MooseEnum>("flux_limiter_type").getEnum<FluxLimiterTypeEnum>()),
+    _nodes_to_elem_map(0)
+{
+  // TODO: test this
+  if (!_execute_enum.contains(EXEC_TIMESTEP_BEGIN) || !_execute_enum.contains(EXEC_NONLINEAR))
+    mooseError("The AdvectiveFluxCalculator UserObject " + name() +
+               " execute_on parameter must include, at least, 'timestep_begin nonlinear'");
+}
+
+void
+AdvectiveFluxCalculator::timestepSetup()
+{
+  if (_nodes_to_elem_map.size() == 0)
+  {
+    // QUERY: We don't want to recompute the neighboring nodes every nonlinear iteration.  We
+    // only need to do that at the beginning and when the mesh has changed because of adaptivity
+    // (and anything else i haven't thought about?)
+    // QUERY: does this properly account for multiple processors?
+    // QUERY: or do we want to use _mesh.nodeToElemMap() ?   (if so, then
```

Comment:
Yes, this is already in MooseMesh, so use that instead of solving that problem again

Reply:
I can use `nodeToElemMap` (which returns a `std::map< dof_id_type, std::vector< dof_id_type > >` )but then i can't see how i can use `MeshTools::find_nodal_neighbors` since the latter requires a `std::vector<std::vector<const Elem *>>`.  I don't really care about finding neighbors next to nodes: i just want to find the nodal neighbors, but the elements seem to be a necessary intermediate step

Reply:
I misunderstood line 74 and 75 below. I thought you were trying to build what we already have in MooseMesh. On second look, it is not the case, so your code is fine.

But, you will want to change the local variable name from `_nodes_to_elem_map` to `nodes_to_elem_map` :-)